### ISSUE TITLE
[Android] React native plans, and/or permission to work on it? 

### ISSUE THREAD

**Issue by williamconnolly** [200ok-ch/organice#251]
**Is your feature request related to a problem? Please describe.**
I love how slick the android webapp in Chrome is, but there are quite a few quality-of-life improvements that would benefit from a native solution. A sizeable amount of code reuse too, in this project.

**Describe the solution you'd like**
To hear about plans that have been made, suggested in the past, or active, and request to help work on it. 

**Describe alternatives you've considered**
I couldn't find any mentions of React native in the PRs or Issues.

**Comment by munen** [comment]
Hi @williamconnolly 

Thank you for the kind words on organice and stepping up with ambitions to contribute to it!

I do have plans to make organice a hybrid app at some point. The most compelling reason for me is to integrate notifications for scheduled headers which is currently not possible within the scope of a PWA. However, it is possible to do it from a browser, so chances are not too bad that it might come to Android and iOS at some point as well.

Out of curiousness: Which other 'quality-of-life improvements' were you thinking of?

As for the implementation, I would hold off with a port to React Native. Otherwise we'll quickly have a lot of drift. There's various discussions ([1](https://github.com/200ok-ch/organice/issues/188), [2](https://github.com/200ok-ch/organice/issues/182), [3](https://github.com/200ok-ch/organice/issues/211), etc) that'll have a big impact on the React components.

To not fragment the community in this early stage, I would personally go for a framework where we have a 100% re-use like Cordova. Also, the only use-case for the hybrid app, that I'm thinking of are notifications. Hence, chosing a framework where there's more refactoring required (like RN) wouldn't yield a benefit. Having said that, you might be thinking about other improvements(;
 

TL;DR: I'm not saying this is generally the platform that I prefer for hybrid apps, I have good experiences with React Native. But for this specific case, I'd go with a 'native web container' based framework like Cordova for the aforementioned reasons. If you wanted to work on this and create a PR against this repo, I'd be stoked and be happy to help out!

**Comment by schoettl** [comment]
I agree, @munen.

> Also, the only use-case for the hybrid app, that I'm thinking of are notifications.

Isn't the a) random-killing of service workers and b) offline support another important reason for a hybrid app? Or can't a hybrid app fix this?

Regarding a) and b): I'm on Android and when the service worker is killed and I have no internet, I cannot use the app.

**Comment by munen** [comment]
@schoettl I agree that complete offline support is a good goal! However, it's not required to do a full rewrite in a different stack to go fully offline^^

Building a completely offline PWA has a couple extra steps - just as a fully offline native or hybrid app would have. There's just a little more work to be done to support offline capabilities further (here's a good resource to start: https://developers.google.com/web/fundamentals/instant-and-offline/web-storage/offline-for-pwa).

**Comment by williamconnolly** [comment]
Thanks for the prompt reply! 

I agree regarding rewriting to React native at this point in time. Code drift is significant when the largest category of code in this project is components/JSX code. This would all need to be ported - so I think that I will try to focus my efforts on quality-of-life improvements as a PWA on Android.

• Tap toggling the "edit content" on mobile for a heading, especially when it is the bottom-most list element on the screen, should scroll the element up enough that the content is visible. Similar with scrolling back down when its hidden. 
• More-easily-accessible undo/redo functionality - this is a tough one to figure out how to solve for both mobile platforms. I wish they had better OS-wide patterns for such a basic concept.
• Heading TODO cycling support that doesn't rely on tapping the TODO itself -- my setup has TODO/DONE/BLANK and so once it is is blank mode, I cannot toggle back to todo. 

The last one is not mobile specific. Sounds like I should accept and work around the mobile browser limitations, and get to work being the change I want to see in the world!

**Comment by munen** [comment]
> Thanks for the prompt reply! 

Sure thing^^

> I agree regarding rewriting to React native at this point in time.
> Code drift is significant when the largest category of code in this
> project is components/JSX code. This would all need to be ported -
> so I think that I will try to focus my efforts on quality-of-life
> improvements as a PWA on Android.

👍

> • Tap toggling the "edit content" on mobile for a heading, especially when it is the bottom-most list element on the screen, should scroll the element up enough that the content is visible. Similar with scrolling back down when its hidden. 

I agree. That would be a great improvement for Android and iOS. It'll be a little tricky to build/test it for devices with different form factors/resolutions. I can happily test and propose changes for iOS where I'm usually testing on an iPhone Xs and 6s for the two most common form factors.

> • More-easily-accessible undo/redo functionality - this is a tough one to figure out how to solve for both mobile platforms. I wish they had better OS-wide patterns for such a basic concept.

FYI, for the future we're looking to bring more and more [Material Design](https://material.io/) into organice. Maybe that helps you thinking about how to approach this topic.

In any case, I'm looking forward on your take on it! I agree that reaching for undo/redo us not easily done in a bigger file when you're drilled into it. That's not great for an important feature.

> • Heading TODO cycling support that doesn't rely on tapping the TODO itself -- my setup has TODO/DONE/BLANK and so once it is is blank mode, I cannot toggle back to todo. 

I have good news for you! "Tap TODO to advance state" is an optional feature that you can enable in the settings. The _default_ feature is that you can advance state as would be more common in a mobile app - by swiping 🎊

Swiping left will delete a todo. Swiping right will toggle the TODO state as in Emacs - hence a heading without a TODO will be brought into the first state of the sequence (TODO in your case).

The docs for such things are in the sample.org file (accessible in the [settings](https://organice.200ok.ch/sample)), or on Github here: https://github.com/200ok-ch/organice/blob/master/sample.org#todos-14-25

> The last one is not mobile specific. Sounds like I should accept and work around the mobile browser limitations, and get to work being the change I want to see in the world!

🙏

**Comment by munen** [comment]
Since we are in agreement regarding the rough short-term roadmap and you already suggested specific PWA features to build instead of working on RN for the time being, I suggest we close this issue and either create three new issues for the features you mentioned and assign you to them. Or you can just bring a PR whenever you're ready - whatever works best for you.

Is that OK for you?

**Comment by williamconnolly** [comment]
I've taken notes locally on the features I'd like to focus on, and proposed feedback/solutions by you two, so I'll bring a PR when ready! Otherwise it just clutters the issue list. 

Closing this issue.
